polkadot/runtime/common/src/paras_registrar/migration.rs

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use frame_support::{
+	dispatch::GetStorageVersion,
+	traits::{Contains, OnRuntimeUpgrade, StorageVersion},
+};
+
+#[derive(Encode, Decode)]
+pub struct ParaInfoV1<Account, Balance> {
+	manager: Account,
+	deposit: Balance,
+	locked: bool,
+}
+
+pub struct MigrateToV1<T, UnlockParaIds>(sp_std::marker::PhantomData<(T, UnlockParaIds)>);
+impl<T: Config, UnlockParaIds: Contains<ParaId>> OnRuntimeUpgrade
+	for MigrateToV1<T, UnlockParaIds>
+{
+	fn on_runtime_upgrade() -> Weight {
+		let onchain_version = Pallet::<T>::on_chain_storage_version();
+
+		if onchain_version == 0 {
+			let mut count = 1u64; // storage version read write
+			Paras::<T>::translate::<ParaInfoV1<T::AccountId, BalanceOf<T>>, _>(|key, v1| {
```

Comment:
`translate` can silently fail on decode. Can you please rather use iterators/for loop?

Reply:
If we shouldn't use translate, maybe we should either remove it or fix it?
But what else can I do on failed decode with loop?

Reply:
Yes there is no real way to handle this since the migration is infallible, but doing a `defensive!` so it will show up in tests and log an error on the real run would be better I think.

Reply:
so we should update `translate` to use `defensive!`